### DISCUSSION TITLE
C64 multicolor mode patterns and colors

### DISCUSSION THREAD

**Discussion by coslor** [drmortalwombat/oscar64#49] [Q&A]
I've been going through the source, but the light bulb just isn't coming on -- what exactly does the 8-bit "color" param do in calls like: `bmmcu_rect_fill()` and `bmmcu_line()`? I've been experimenting, and it looks like some bits determine which color to use out of the defined multicolor registers, and some different bit patterns, but it's not clear what's what. 

Does anybody have the details of exactly what this param does?

P.S. love oscar64 so far! I've come across a few bugs here & there, but it generally produces great (and pretty speedy!)  6502 code. Thanks for sharing this!

**Comment by drmortalwombat** [comment]
The color parameter is just two bit, so the four possible color values of multicolor mode.  

Pattern is an array of 8 bytes that is used to fill the color 8x8 cell.  So the first byte of the pattern will be used for each eigth row.  With multicolor mode the pattern for a single row (one byte) consists of four 2bit color values.

**Reply by coslor** [reply]
Thanks for the quick response! That makes sense, but the functions like `bmmcu_line()` and `bmmcu_rect_fill()` in mcbitmap don't ask for separate color and pattern arguments -- they just take `color` which seems to define both the color used and the pattern used (3, for example, appears to be color `vic.color_back` and a solid pattern, while 7 is `vic.color_back` and a dithered pattern). 

Is it just that the lower 2 bits are the color and the upper 6 bits the pattern? I get what seem like inconsistent results while trying to use it like that, but I could just be doing something dumb. 

Is there a list of what the different multicolor patterns are (vertical lines, 4 dots, tight dither pattern, etc.)? Or is it enumerated somewhere in the source?

**Reply by drmortalwombat** [reply]
bmmcu_line and bmmcu_rect_fill both work with a solid color from 0 to 3

**Comment by coslor** [comment]
Great, thanks!
